Create a blank FAT32 image

DrDeano opened the pull request:
This will be used for testing the FAT32 driver for the kernel and will be integrated into the OS as a mkfs.fat32 program.
Plus typos

Part of #216

SamTebbs33 commented on build.zig:
```diff
@@ -28,20 +30,24 @@ pub fn build(b: *Builder) !void {
 
     const main_src = "src/kernel/kmain.zig";
     const arch_root = "src/kernel/arch";
+    const fat32_bootstrap_src = "fat32_bootstrap.asm";
```
What's the purpose of the fat32_bootstrap?

DrDeano replied:
It is part of the FAT32 header. Is it used if it is loaded as a bootloader and will just print "please insert a bootable media, press any key to restart." It is taken from my bootloader. I guess it isn't 100% needed, but it part of the specification so may not be able to mount in Linux. There could just be a random flat binary, but the code must come from somewhere.

SamTebbs33 replied:
That makes sense. I think it would be nice to move it somewhere under arch/x86 though.

DrDeano replied:
It shouldn't go into the x86 folder as it can be used for any arch. Maybe could add the pre-build binary to the git repo and remove the build step for this, so removing the nasm dependency. And if people want to build it, they can separetly

SamTebbs33 replied:
Building the image separately makes a lot of sense 👍 Building a FAT image shouldn't be a dependency for the kernel. Although, I don't see how this can be used for any architecture since it uses x86 instructions.

SamTebbs33 commented on build.zig:
```diff
@@ -166,6 +178,226 @@ pub fn build(b: *Builder) !void {
+    builder: *Builder,
+
+    /// The file path to the bootstrap compiled code to be used in the FAT32 header.
+    fat32_bootstrap_bin: []const u8,
+
+    /// The path to where the ramdisk will be written to.
+    out_file_path: []const u8,
+
+    /// Options for creating the FAt32 image.
+    fat32_options: Options,
```
I think the `fat32_` prefixes here is unnecessary.

SamTebbs33 commented on build.zig:
```diff
@@ -166,6 +178,226 @@ pub fn build(b: *Builder) !void {
+
+        /// The FAT32 image size in bytes. This is not the formatted size. Default see
+        /// `Fat32.getDefaultImageSize`
+        image_size: DefaultOrValue = .default,
```
Instead of setting these to `.default`, could they just be set to the default value? Feel free to explain the reason why there's an extra level of indirection.

DrDeano replied:
That is a good point lol. I just went with it because I kinda like it ;)

SamTebbs33 commented on mkfat32.zig:
```diff
@@ -0,0 +1,361 @@
+    /// parameter block for FAT32 with the extended signature 0x29.
+    pub const Header = struct {
+        /// The OEM.
+        oem: [8]u8 = "ZystemOS".*,
```
It's almost as if we planned the name to be the exact length needed :D

DrDeano replied:
I know right :P

SamTebbs33 commented on mkfat32.zig:
```diff
@@ -0,0 +1,361 @@
+        try image.seekableStream().seekTo(fat32_header.fsinfo_sector * fat32_header.bytes_per_sector);
+
+        // First signature
+        try image.writer().writeIntLittle(u32, 0x41615252);
```
Can `image.writer()` be stored in a variable instead of being called every time?

DrDeano replied:
In the std lib, when multiple writes are made, `writer()` is called each time. But it didn't make a difference when storing the write in a variable. ??

SamTebbs33 replied:
I just think it would be a little cleaner.

DrDeano replied:
ok

SamTebbs33 commented on mkfat32.zig:
```diff
@@ -0,0 +1,361 @@
+    /// Return: u32
+    ///     Serial number for a new FAT32 image.
+    ///
+    pub fn createSerialNumber(current_time: SerialTime) u32 {
```
This could be moved to the SerialTime struct for nice encapsulation.

DrDeano replied:
I don't think it should be moved as it is a static function and there isn't a need for `Self` for `SerialTime`. It would encapsulate it but would make the use of it longer: `Fat32.createSerialNumber()` =>`Fat32.SerialTime.createSerialNumber()`. Also will need to make the struct public.

SamTebbs33 replied:
It's static for now, but with my suggestion, if you have a `SerialTime` instance you'd just do `serial_time.createSerialNumber()` instead of `Fat32.createSerialNumber(serial_time)`, so it would actually be shorter!

DrDeano replied:
There wouldn't be any other need for SerialTime other than creating the serial number. The only reason I have the SerialTime struct was to reduce the number of parameters being passed to the `createSerialNumber`. It would just create code blot and be more complex than needed. Once there is std lib get human readable time, I can have another look at changing how this is done. Probable removing the the argument for `createSerialNumber`, removing the `SerialTime` struct, and having the function call the std lib function directly.

SamTebbs33 replied:
Hmm, I still think that putting the function in the struct is better than having it in Fat32, since that's how we and the std lib do things in other structs. I don't agree with the code bloat argument since it won't mean adding more lines and it would actually be cleaner! Alternatively I don't mind having 6 parameters for the function.

DrDeano replied:
I shall do parameters

SamTebbs33 reviewed (approved):
LGTM

DrDeano committed:
Create a blank FAT32 image

This will be used for testing the FAT32 driver for the kernel and will be integrated into the OS as a mkfs.fat32 program.
Plus typos

Fixed dependencies


Removed `fat32_` in options

Plus fixed doc comment

Removed the DefaultOrValue

Also reordered some stuff

Removed the serial time for more parameters


Moved writer() and seekableStream() to variables


Refactored mkFAT32
